File: 4-inheritance/new.c
#include <malloc.h>
#include <stdarg.h>
#include <assert.h>

#include "new.r"
#include "new.h"


void * new (const void * _class, ...) {
    const struct Class * class = _class;

    void * item = calloc(1, class -> size);
    assert(item);

    * (const struct Class **) item = class;
    
    if (class -> ctor) {
        va_list ap;

        va_start(ap, _class);
        item = class -> ctor(item, & ap);
        va_end(ap);
    }

    return item;
}

void delete (void * _self) {
    const struct Class ** cp = _self;

    if (_self && * cp && (* cp) -> dtor) {
        _self = (* cp) -> dtor(_self);
    }

    free(_self);
}

void draw (const void * self) {
    const struct Class * const * cp = self;
    assert(self && * cp && (* cp) -> draw);

    (* cp) -> draw(self);
}

File: 4-inheritance/new.r
#ifndef CLASS_H
#define CLASS_H

#include <stddef.h>
#include <stdarg.h>


/**
 * Declaration of Class.
 *
 * PRIVATE DEFINITION HEADER
 *
 * It contains:
 *
 * - size:   The length that new() allocates for an object.
 *
 * - ctor:   Points to the constructor called by new() which recieves
 *           the allcated memory and the rest of the argument list
 *           passed to new() originally.
 *
 * - dtor:   Points to the destructor called by delete() which recieves
 *           the object to be destroyed.
 *
 * - clone:  Points to a copy functions which recieves
 *           the object to be copied.
 *
 * - differ: Points to a function which compares its object
 *           to smth else.
 */
struct Class {
    size_t        size;
    void *        (* ctor) (void * self, va_list * app);
    void *        (* dtor) (void * self);
    void          (* draw) (const void * self);
};


#endif        /* CLASS_H */

